cpp/src/strings/convert/convert_fixed_point.cu

```diff
@@ -33,74 +33,152 @@
 #include <rmm/cuda_stream_view.hpp>
 #include <rmm/exec_policy.hpp>
 
+#include <thrust/optional.h>
 #include <thrust/transform.h>
 
 namespace cudf {
 namespace strings {
 namespace detail {
 namespace {
+
+struct string_to_decimal_base {
+  /**
+   * @brief Return the integer component of a decimal string.
+   *
+   * This is reads everything up to the exponent 'e' notation.
+   * The return includes the integer digits and any exponent offset.
+   *
+   * @param[in,out] iter Start of characters to parse
+   * @param[in] end End of characters to parse
+   * @return Integer component and exponent offset.
+   */
+  __device__ thrust::pair<uint64_t, int> parse_integer(char const*& iter,
+                                                       char const* iter_end) const
+  {
+    // highest value where another decimal digit cannot be appended without an overflow;
+    // this preserves the most digits when scaling the final result
+    constexpr uint64_t decimal_max = (std::numeric_limits<uint64_t>::max() - 9L) / 10L;
+
+    uint64_t value     = 0;  // for checking overflow
+    int exp_offset     = 0;
+    bool decimal_found = false;
+
+    while (iter < iter_end) {
+      auto const ch = *iter++;
+      if (ch == '.' && !decimal_found) {
+        decimal_found = true;
+        continue;
+      }
+      if (ch < '0' || ch > '9') {
```

Comment:
Is there a device version of `isdigit`?
```suggestion
      if (not isdigit(ch)) {
```

Reply:
No, there is no device version of `std::isdigit()`.